src/aleph/schemas/pending_messages.py

```diff
@@ -0,0 +1,175 @@
+
+from aleph.exceptions import InvalidMessageError
+from aleph.utils import item_type_from_hash
+
+MAX_INLINE_SIZE = 200000  # 200kb max inline content size.
+
+
+class BasePendingMessage(BaseModel):
+    """
+    A raw Aleph message, as sent by users to the Aleph network.
+    """
+
+    sender: str
+    chain: str
+    signature: str
+    type: MessageType
+    item_content: Optional[str]
+    item_type: ItemType
+    item_hash: str
+    time: float
+    channel: Optional[str] = None
+    content: Optional[BaseContent] = None
+
+    @root_validator(pre=True)
+    def load_content(cls, values):
+        """
+        Preload inline content. We let the CCN populate this field later
+        on for ipfs and storage item types.
+        """
+
+        item_type = values["item_type"]
+        item_content = values.get("item_content")
+        if item_type == ItemType.inline:
+            if item_content is None:
+                raise ValueError("Item content not specified for inline item type")
+
+            if len(item_content) > MAX_INLINE_SIZE:
+                raise ValueError("Message too long")
+            try:
+                values["content"] = json.loads(item_content)
+            except json.JSONDecodeError as e:
+                raise ValueError("Could not load item content") from e
+        else:
+            if item_content is not None:
+                raise ValueError(f"{item_type} messages cannot define item_content")
+
+        return values
+
+    @root_validator()
+    def check_item_type(cls, values):
+        """
+        Checks that the item hash of the message matches the one inferred from the hash.
+        Only applicable to storage/ipfs item types.
+        """
+        item_type = ItemType(values["item_type"])
+        if item_type == ItemType.inline:
+            return values
+
+        expected_item_type = item_type_from_hash(values["item_hash"])
+        if item_type != expected_item_type:
+            raise ValueError(
+                f"Expected {expected_item_type} based on hash but item type is {item_type}."
+            )
+        return values
+
+    @validator("item_hash")
+    def check_item_hash(cls, v, values):
+        """
+        For inline item types, check that the item hash is equal to
+        the hash of the item content.
+        """
+
+        item_type = values["item_type"]
+        if item_type == ItemType.inline:
+            item_content: str = values["item_content"]
+
+            computed_hash: str = sha256(item_content.encode()).hexdigest()
+            if v != computed_hash:
+                raise ValueError(
+                    "'item_hash' do not match 'sha256(item_content)'"
+                    f", expecting {computed_hash}"
+                )
+        elif item_type == ItemType.ipfs:
+            # TODO: CHeck that the hash looks like an IPFS multihash
+            pass
+        else:
+            assert item_type == ItemType.storage
+        return v
```

Comment:
Just tried it, the validator does not get in the way but I removed the `hash_type` member of the message model, if only because we will probably never use it (better to use multihash/multicodec instead). If that's okay for you, I think I'll leave it like that for the moment, and we'll do a large update of aleph_message once we're fixed on the models.

I basically want to have a class tree like this:

* `PendingMessage`: base class, = the data sent by aleph-client or similar
* `DbMessage`: inherits PendingMessage (or a common base class), = the schema of the data stored in MongoDB.
* `ApiMessage`: inherits DbMessage (or a common base class), = the schema of the API response. Similar to the current aleph_message Message.